deps/oculus/src/ovrsession.cpp

```diff
@@ -0,0 +1,507 @@
+#include <deque>
+#include <iostream>
+#include <ovrsession.h>
+#include <stdlib.h>
+#include <mutex>
+#include <node.h>
+#include <thread>
+#include <v8.h>
+
+#include <Extras/OVR_Math.h>
+
+using namespace v8;
+using namespace OVR;
+
+namespace oculusvr {
+  uv_sem_t reqSem;
+  uv_async_t resAsync;
+  std::mutex reqMutex;
+  std::mutex resMutex;
+  std::deque<std::function<void()>> reqCbs;
+  std::deque<std::function<void()>> resCbs;
+  std::thread reqThread;
+
+  void RunResInMainThread(uv_async_t *handle) {
+    Nan::HandleScope scope;
+
+    std::function<void()> resCb;
+    {
+      std::lock_guard<std::mutex> lock(resMutex);
+
+      resCb = resCbs.front();
+      resCbs.pop_front();
+    }
+    if (resCb) {
+      resCb();
+    }
+  }
+};
+
+OculusVRPosRes::OculusVRPosRes(Local<Function> cb) : cb(cb) {}
+
+OculusVRPosRes::~OculusVRPosRes() {}
+
+NAN_MODULE_INIT(OVRSession::Init)
+{
+  // Create a function template that is called in JS to create this wrapper.
+  Local<FunctionTemplate> tpl = Nan::New<FunctionTemplate>(New);
+
+  // Declare human-readable name for this wrapper.
+  tpl->SetClassName(Nan::New("OVRSession").ToLocalChecked());
+
+  // Declare the stored number of fields (just the wrapped C++ object).
+  tpl->InstanceTemplate()->SetInternalFieldCount(1);
+
+  Nan::SetPrototypeMethod(tpl, "GetControllersInputState", GetControllersInputState);
+  Nan::SetPrototypeMethod(tpl, "GetPose", GetPose);
+  Nan::SetPrototypeMethod(tpl, "Submit", Submit);
+  Nan::SetPrototypeMethod(tpl, "SetupSwapChain", SetupSwapChain);
```

Comment:
From my reading this isn't called from JS and swap chains are managed from C++. Maybe we should get rid of this exposure and the `NAN_METHOD` to prevent confusion?

Reply:
Actually, I just saw [this](https://github.com/exokitxr/exokit/pull/787/files#diff-1fdf421c05c1140f6d71444ea2b27638R484). This is confusing since it seems both JS and C++ manage the swap chain.

I think if that's the case we should settle on one code path to do it -- in this case it seems `Oculus_Init` in C++ could set up the first swap chain.